### PR TITLE
Sofar/aadu/usart pipeline

### PR DESCRIPTION
**Pull request by beardedtrogon**
Pipeline now uses DMA for USART RX/TX. Fixed bugs to allow for 255 byte payloads and removed parsing logic from ISR into rx_task

### CODE REVIEW

File: drivers/console/bm_serial.c
Comment (alvarop) [inline drivers/console/bm_serial.c]: Recommend using a `#define NUM_DEVICES 3` (or similar)
If we ever change the number, you'll have to go through all of your code looking for that 3 to replace it, which will be error prone

File: drivers/console/bm_serial.c
Comment (alvarop) [inline drivers/console/bm_serial.c]: nit: indentation

File: drivers/console/bm_serial.c
Comment (alvarop) [inline drivers/console/bm_serial.c:241]: Does this happen in interrupt or task context?

File: drivers/console/bm_serial.c
Comment (alvarop) [reply drivers/console/bm_serial.c]: Actually, looks like you're mixing spaces and tabs 😱

Use whichever you chose, just don't mix them up 😂

File: drivers/console/bm_serial.c
Comment (alvarop) [inline drivers/console/bm_serial.c]: Is `encoded_rx_buf` the same buffer DMA is writing to?
If so, we might consider making a copy of that buffer so DMA can keep doing it's thing ASAP.

I know we're ping-ponging, but if this decoding thread takes too long for whatever reason (other things going on in the system), DMA will happily overwrite the contents of the buffer and you'll end up decoding two different messages (thinking it's one)

File: drivers/console/bm_serial.c
Comment (alvarop) [inline drivers/console/bm_serial.c]: 😂

File: drivers/console/bm_serial.c
Comment (alvarop) [inline drivers/console/bm_serial.c]: Should we check that the message length is less than the maximum somewhere? Don't want to overflow `tx_enc_buf`!

File: drivers/console/bm_serial.c
Comment (alvarop) [inline drivers/console/bm_serial.c]: Same here, should we check that frame_length+crc size <= size of the `tx_payload_buf`

**Review by alvarop — COMMENTED** [review]
Haven't gone through `drivers/serial/uart_bm.c`, but left comments on bm_serial.c

File: drivers/console/bm_serial.c
Comment (beardedtrogon) [reply drivers/console/bm_serial.c:241]: This should happen within the ISR

File: drivers/console/bm_serial.c
Comment (beardedtrogon) [reply drivers/console/bm_serial.c]: No, the buffers that DMA write to are dev_ctx[i].buf (1 for each serial interface). In the cb function, the ISR memcpys the value from these buffers into the encoded_rx_buf ping pong buffer.

File: drivers/serial/CMakeLists.txt
Comment (spiderkeys) [inline drivers/serial/CMakeLists.txt]: Worth thinking about renaming this series of files to uart_stm32_bm, to both show that parts of it originated out of the uart_stm32 driver and also that it only works on STM32.

File: dts/bindings/serial/st,stm32-uart-bm.yaml
Comment (spiderkeys) [inline dts/bindings/serial/st,stm32-uart-bm.yaml]: I think this single-wire mode and any functionality related to it could/should be removed, since our driver and PHY are not designed to support this modality.

Not a big deal if it ends up being more trouble than it is worth right now.

File: dts/bindings/serial/st,stm32-usart-bm.yaml
Comment (spiderkeys) [inline dts/bindings/serial/st,stm32-usart-bm.yaml]: Same comment as uart yaml

File: samples/net/bm_serial_test/prj.conf
Comment (spiderkeys) [inline samples/net/bm_serial_test/prj.conf]: How about we update the Kconfig files and/or driver to make it so that UART_ASYNC_API is either no longer required (ideal) or such that selecting UART_BM automatically selects UART_ASYNC_API. Since uart_bm depends on DMA and the async events, its use should imply the use of any required dependencies.

Comment (spiderkeys) [reply samples/net/bm_serial_test/prj.conf]: This would likely be done in drivers/serial/Kconfig.bm

File: drivers/serial/uart_bm.h
Comment (spiderkeys) [inline drivers/serial/uart_bm.h]: Driver always requires interrupts/async mode, so these can be dropped

File: drivers/serial/uart_bm.h
Comment (spiderkeys) [inline drivers/serial/uart_bm.h]: single_wire could be removed, as suggested in the device tree yaml.

File: drivers/serial/uart_bm.h
Comment (spiderkeys) [inline drivers/serial/uart_bm.h]: Given that we have only two wires to communicate over, I'm guessing that we can probably cut out hardware flow control support directly.

File: drivers/serial/uart_bm.h
Comment (spiderkeys) [inline drivers/serial/uart_bm.h]: ifdef can be removed if mandatory dependency on async api specified

File: drivers/serial/uart_bm.h
Comment (spiderkeys) [inline drivers/serial/uart_bm.h]: This should probably be renamed to uart_bm_dma_stream, to prevent potential conflicts with the original uart_stm32 driver.

File: drivers/serial/uart_bm.h
Comment (spiderkeys) [inline drivers/serial/uart_bm.h]: ifdef removable

File: drivers/serial/uart_bm.h
Comment (spiderkeys) [inline drivers/serial/uart_bm.h]: ifdef removable

File: drivers/serial/uart_bm.c
Comment (spiderkeys) [inline drivers/serial/uart_bm.c]: Might want to consider adding Sofar to Copyright

File: drivers/serial/uart_bm.c
Comment (spiderkeys) [inline drivers/serial/uart_bm.c]: Debugging GPIO should probably be removed before merging this

File: drivers/serial/uart_bm.c
Comment (spiderkeys) [inline drivers/serial/uart_bm.c]: Potentially removable API - default to hw flow control disabled

File: drivers/serial/uart_bm.c
Comment (spiderkeys) [inline drivers/serial/uart_bm.c]: Potentially removable API - default to 8 data bit width. I don't see us ever changing this while using a UART as the communication peripheral.

File: drivers/serial/uart_bm.c
Comment (spiderkeys) [inline drivers/serial/uart_bm.c]: Polling APIs removable

File: drivers/serial/uart_bm.c
Comment (spiderkeys) [inline drivers/serial/uart_bm.c]: FIFO APIs removable - Can set .fifo_fill and .fifo_read methods to NULL.

File: drivers/serial/uart_bm.c
Comment (spiderkeys) [inline drivers/serial/uart_bm.c]: Removable ifdef - mandatory requirement

File: drivers/serial/uart_bm.c
Comment (spiderkeys) [inline drivers/serial/uart_bm.c]: Debug log statements can be removed from these functions: these can be registered separately in the user_callback, if desired.

File: drivers/serial/uart_bm.c
Comment (spiderkeys) [inline drivers/serial/uart_bm.c]: Remove debug GPIO code

File: drivers/console/bm_serial.c
Comment (spiderkeys) [inline drivers/console/bm_serial.c:15]: ^ Comments should be updated to remove mention of UART Pipe and add a description for bm_serial. Probably should add Sofar to the copyright - this driver has substantially diverged from the original uart pipe code.

File: include/drivers/console/bm_serial.h
Comment (spiderkeys) [inline include/drivers/console/bm_serial.h]: Some of these defines should get converted to Kconfig options so they can be easily set via prj.conf. The following seem like good candidates:
- NUM_BM_FRAMES
- MAX_SERIAL_DEV_COUNT
- TASK_STACK_SIZE

File: include/drivers/console/bm_serial.h
Comment (spiderkeys) [inline include/drivers/console/bm_serial.h]: NUM_DEC_MSGS doesn't seem to be used anywhere.

File: samples/net/bm_serial_test/prj.conf
Comment (beardedtrogon) [reply samples/net/bm_serial_test/prj.conf]: It looks like certain structs defined outside the scope of this driver depend on the UART_ASYNC_API being set. I can go with the latter approach of automatically selecting this configuraiton

File: drivers/console/bm_serial.c
Comment (spiderkeys) [inline drivers/console/bm_serial.c]: Pretty much all of the of the globals in this file, as far as I can tell, are referenced in ISRs or are used in multiple threads, and should be marked as volatile.

Similarly, all globals defined here are only used in this .c file, so should be marked static to avoid potential multiple definition issues in the future.

**Review by spiderkeys — CHANGES_REQUESTED** [review]
Approved, pending the static/volatile changes.

**Pull request merged**
